### PR TITLE
[quant][fx] Move all fusion registrations to backend_config_dict

### PR DESCRIPTION
**Pull request by jerryzh168**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #75521
* #75520
* #75469
* #75389
* #75388
* #75387
* #75386
* #75378
* #75377
* __->__ #75318

Summary:
This PR moves the registrations for fusion patterns to backend_config_dict

Also fixed one issue in numeric suite graph matcher, since now (torch.nn.ReLU, torch.nn.BatchNorm3d)
would appear in quant patterns, (previously only in fusion pattern), and we need to match sure (torch.nn.ReLU, (torch.nn.BatchNorm3d, torch.nn.Conv3d))
can match before (torch.nn.ReLU, torch.nn.BatchNorm3d), but previously, it looks like (torch.nn.ReLU, (torch.nn.BatchNorm3d, torch.nn.Conv3d)) is not
really matched since `end_node_matches_reversed_fusion` is expecting a flattened pattern like (torch.nn.ReLU, torch.nn.BatchNorm3d, torch.nn.Conv3d),
for now we'll manually flatten this pattern, but in the future I think we might want to use the matching function `is_match` under torch.ao.quantization.fx.match_utils
to do this matching.

Test Plan:
python test/test_quantization.py TestQuantizeFx
python test/test_quantization.py TestQuantizeFxOps
python test/test_quantization.py TestFXNumericSuiteCoreAPIs

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D35423788](https://our.internmc.facebook.com/intern/diff/D35423788)

### CODE REVIEW

**Comment by jerryzh168** [comment]
@jerryzh168 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D35423788).

File: torch/ao/quantization/fx/backend_config/native.py [hunk 1]
```diff
@@ -399,27 +467,25 @@ def _get_bn_configs():
         torch.nn.BatchNorm2d: nni.BNReLU2d,
         torch.nn.BatchNorm3d: nni.BNReLU3d,
     }
-    for bn in [torch.nn.BatchNorm2d, torch.nn.BatchNorm3d]:
+    for bn in bn_to_fused_bn.keys():
         # TODO: enable these and remove entries in fusion_patterns.py
```
Comment (vkuzo) [inline torch/ao/quantization/fx/backend_config/native.py:471]: is this comment still valid?

**Review by vkuzo — APPROVED** [review]

**Review by andrewor14 — APPROVED** [review]

File: torch/ao/quantization/fx/backend_config/native.py (hunk 1 above)
Comment (jerryzh168) [reply torch/ao/quantization/fx/backend_config/native.py:471]: no, will remove

File: torch/ao/quantization/fx/backend_config/native.py [hunk 2]
```diff
@@ -146,6 +161,19 @@ def _get_linear_configs():
...
+    linear_configs.append({
+        "pattern": (torch.nn.ReLU, torch.nn.Linear),
+        "dtype_configs": dtype_configs ,
+        "fuser_method": reverse_sequential_wrapper2(nni.LinearReLU),
+    })
+    linear_configs.append({
+        "pattern": (torch.nn.functional.relu, torch.nn.Linear),
+        "dtype_configs": dtype_configs ,
+        "fuser_method": reverse_sequential_wrapper2(nni.LinearReLU),
+    })
```
[the remark this answers is not in the document]
Comment (jerryzh168) [reply torch/ao/quantization/fx/backend_config/native.py:174]: oh I see, yeah I can remove that one

File: torch/ao/quantization/fx/backend_config/native.py [hunk 3]
```diff
@@ -192,6 +220,14 @@ def _get_linear_configs():
...
     # (3) Linear + batchnorm
     # ------------------------
+    # 3.1 linear bn fusion
+    linear_configs.append({
+        "pattern": (nn.BatchNorm1d, nn.Linear),
+        "dtype_configs": dtype_configs ,
+        "fuser_method": reverse2(fuse_linear_bn)
+    })
+
+    # 3.2 linear bn quantization
```
[the remark this answers is not in the document]
Comment (jerryzh168) [reply torch/ao/quantization/fx/backend_config/native.py:230]: this config is used in fusion, after this is fused, the pattern won't appear in the following qunatization flow, and theoretically we could just remove this config

[the remark this answers is not in the document]
Comment (jerryzh168) [reply torch/ao/quantization/fx/backend_config/native.py:230]: depends on what we mean by quantization I guess, if quantization = fusion + insert observers + convert then yes, both are used in quantization.
if quantization = insert observers + convert then fusion and quantization are separate steps.

Maybe we could have a different name for `insert observers + convert` to avoid overloading "quantization"

File: torch/ao/quantization/fx/backend_config/native.py [hunk 4]
```diff
@@ -294,6 +330,30 @@ def _get_conv_configs():
...
+        # TODO: enable when fuse method supports F.relu
+        # conv_configs.append({
+        #     "pattern": (F.relu, (convs.bn, convs.root)),
+        #     "dtype_configs": dtype_configs,
+        #     "root_module": convs.root,
+        #     "fuser_method": reverse3(fuse_conv_bn_relu),
+        # })
+        # TODO: we can add fusion for torch.relu as well
+
+        # 3.2 conv + bn (+ relu) quantization
```
[the remark this answers is not in the document]
Comment (jerryzh168) [reply torch/ao/quantization/fx/backend_config/native.py:356]: this is fused conv + bn (+ relu) quantization config, let me add "fused" to make it more clear

File: torch/ao/quantization/fx/backend_config/native.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (jerryzh168) [reply torch/ao/quantization/fx/backend_config/native.py:230]: fusion/fused sounds good I feel, I will make this change

[the remark this answers is not in the document]
Comment (jerryzh168) [reply torch/ao/quantization/fx/backend_config/native.py:230]: fusion configs will be needed, I don't think we can remove them, they are used in the fuse.py function

**Comment by jerryzh168** [comment]
@jerryzh168 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D35423788).

**Comment by jerryzh168** [comment]
@jerryzh168 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D35423788).

**Comment by jerryzh168** [comment]
@jerryzh168 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D35423788).

**Comment by jerryzh168** [comment]
@jerryzh168 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D35423788).
